src/RA_RichPresence.cpp

```diff
@@ -1,266 +1,260 @@
-                    m_conditionalDisplayStrings.push_back(conditionalString);
-
-                    _ReadTil(EndLine, buffer, 4096, &nCharsRead, pFile);
-                    pBuf = &buffer[0];
-                    m_sDisplay = _ReadStringTil('\n', pBuf, TRUE);
+                    part.m_pLookup = &(*lookup);
+                    part.m_memValue.ParseFromString(sMemValue.c_str());
+                    break;
                 }
             }
-
-            _ReadTil(EndLine, buffer, 4096, &nCharsRead, pFile);
         }
+    } while (true);
+}
 
-        fclose(pFile);
-    }
+bool RA_RichPresenceInterpreter::DisplayString::Test()
+{
+    bool bDirtyConditions, bResetRead; // for HitCounts - not supported in RichPresence
+    return m_conditions.Test(bDirtyConditions, bResetRead);
 }
 
-const std::string RA_RichPresenceInterpreter::Lookup(const std::string& sName, const std::string& sMemString) const
+std::string RA_RichPresenceInterpreter::DisplayString::GetDisplayString() const
 {
-    //	check lookups
-    for (size_t i = 0; i < m_lookups.size(); ++i)
+    std::string sResult;
+    for (const auto& part : m_vParts)
     {
-        if (m_lookups.at(i).Description().compare(sName) == 0)
+        if (!part.m_sDisplayString.empty())
+            sResult.append(part.m_sDisplayString);
+
+        if (!part.m_memValue.IsEmpty())
         {
-            MemValue nValue;
-            nValue.ParseFromString(sMemString.c_str());
-            return m_lookups.at(i).Lookup(static_cast<ra::DataPos>(nValue.GetValue()));
-        }
-    }
+            unsigned int nValue = part.m_memValue.GetValue();
 
-    //	check formatters
-    auto iter = m_formats.find(sName);
-    if (iter != m_formats.end())
-    {
-        MemValue nValue;
-        nValue.ParseFromString(sMemString.c_str());
-        return nValue.GetFormattedValue(iter->second);
+            if (part.m_pLookup != nullptr)
+                sResult.append(part.m_pLookup->GetText(nValue));
+            else
+                sResult.append(MemValue::FormatValue(nValue, part.m_nFormat));
+        }
     }
 
-    return "";
-}
-
-bool RA_RichPresenceInterpreter::Enabled() const
-{
-    return !m_sDisplay.empty();
+    return sResult;
 }
 
-const std::string& RA_RichPresenceInterpreter::GetRichPresenceString()
+static bool GetLine(std::stringstream& stream, std::string& sLine)
 {
-    static std::string sReturnVal;
-    sReturnVal.clear();
+    if (!std::getline(stream, sLine, '\n'))
+        return false;
 
-    if (g_pCurrentGameData->GetGameID() == 0)
+    if (!sLine.empty())
     {
-        sReturnVal = "No game loaded";
-        return sReturnVal;
-    }
-
-    bool bParsingLookupName = false;
-    bool bParsingLookupContent = false;
-
-    std::string sLookupName;
-    std::string sLookupValue;
+        size_t index = sLine.find("//");
+        if (index != std::string::npos)
+        {
+            // if a comment marker was found, remove it and any trailing whitespace
```

Comment:
If comments can be inserted in the middle of a line, an escape character is required in order to be able to display the marker.

Reply:
`//` style comments can't appear in the middle of a line. It is very unlikely that a `//` literal would be desired.

Reply:
From what I can tell, any comment will be stripped from the line before it is interpreted, and comment markers are searched using string::find(), so comments can be inserted in the middle of a line. In fact, your example in this PR's description includes one. Is this incorrect?

Regardless of how likely one is to want to display `//` as part of a script, I don't think it's fair to remove that possibility. For this version I think it would be best to only strip comments found at the beginning of a line, and restore support for comments that begin in the middle of a line along with some escape sequence.

Reply:
From what I can tell, any comment will be stripped from the line before it is interpreted, and comment markers are searched using string::find(), so comments can be inserted in the middle of a line. In fact, your example in this PR's description includes one. Is this incorrect?

Regardless of how likely one is to want to display `//` as part of a script, I don't think it's fair to remove that possibility. For this version I think it would be best to only strip comments found at the beginning of a line, and restore support for comments that begin in the middle of a line along with some escape sequence.

Reply:
You're right, `//` in the middle of a line acts as a line terminator.

The "`@`" marker for a lookup/value in a display string is just as destructive. If the interpreter sees that it looks for an open and close parenthesis to identify the parameter to the lookup:
```
The hero is @ here.
```
Will always generate "The hero is " because the lookup definition is considered incomplete.
```
The hero is @ @Location(0x1234).
```
Will also generate "The hero is ." because it sees " @Location" as the name of the lookup, not "Location". And an undefined lookup generates an empty string.

`//` can be worked around by adding an invalid lookup: `/@()/`. To work around the `@` sign, you have to define a lookup that always evaluates to `@`:
```
Lookup:At
*=@

Display:
The hero is @At(0x0000) here.
```
This was even harder before this PR as you'd have to find a memory value that never changed so the lookup would map to the `@` entry.

Display strings are really the only time escaping would be necessary. I'll modify `GetLine` to ignore a `//` immediately following a `\` and modify `InitializeParts` to treat any character following a `\` literally.

Reply:
Sounds good, thanks.